### ISSUE TITLE
Encoding of 0.0 and -0.0 is not at minimum precision

### ISSUE THREAD

**Issue by winitzki** [peteroupc/CBOR-Java#24]
I'm using CBOR-Java to encode floating-point numbers. I see that the code makes an effort to encode floating-point numbers in the minimum precision possible. For example, `1.0` is encoded in 3 bytes. However, zero values are encoded in 5 bytes. The same happens with floating-precision or double-precision zero values, and also with negative zero values.

```java
assert(CBORObject.FromObject(1.0f).EncodeToBytes().length == 3) // Passes.
assert(CBORObject.FromObject(1.0).EncodeToBytes().length == 3) // Passes.
assert(CBORObject.FromObject(0.0f).EncodeToBytes().length == 3) // Fails. The actual length is 5.
assert(CBORObject.FromObject(-0.0f).EncodeToBytes().length == 3) // Fails. The actual length is 5.
assert(CBORObject.FromObject(0.0).EncodeToBytes().length == 3) // Fails. The actual length is 5.
assert(CBORObject.FromObject(-0.0).EncodeToBytes().length == 3) // Fails. The actual length is 5.
```

The option `float64` is `false` per default. So, I expect minimum-precision encoding but I do not obtain it.

I tracked this to the code of `DoubleToHalfPrecisionIfSameValue()` here: https://github.com/peteroupc/CBOR-Java/blob/master/src/main/java/com/upokecenter/cbor/CBORUtilities.java#L1516-L1527

Zero values will not be detected by this function because, for zero values, both the `exponent` and the `mantissa` are equal to zero. 
https://docs.oracle.com/cd/E19957-01/806-3568/ncg_math.html

So, the value `sexp` in https://github.com/peteroupc/CBOR-Java/blob/master/src/main/java/com/upokecenter/cbor/CBORUtilities.java#L1514 will be computed as `-1008`. Then line https://github.com/peteroupc/CBOR-Java/blob/master/src/main/java/com/upokecenter/cbor/CBORUtilities.java#L1522 detects "underflow" and returns `-1`.

It also seems that the code in https://github.com/peteroupc/CBOR-Java/blob/master/src/main/java/com/upokecenter/cbor/CBORUtilities.java#L1527-L1532 does not work correctly with zero mantissa (it will always return nonzero values even if the mantissa is zero and value `sexp` is in the range for that code to activate).

So, to fix this, I propose to add another `else if` clause explicitly detecting a signed zero value.

I need to encode zero and negative zero values at minimum precision. Is there a workaround? I was not able to see how I could override the non-minimal encoding for those values.

I also tried `CBORObject.FromFloatingPointBits(0L, 2)` to represent floating-point zero at minimum precision, but this reverts to non-minimal encoding because the minimum precision is not detected. I tried other values of "floating point bits" to trick `DoubleToHalfPrecisionIfSameValue()` into accepting the zero value, but so far without any success.

**Comment by peteroupc** [comment]
It can be worked around by returning the correct minimal form for encoding floating-point zeros, either `{0xf9,0,0}` for 0.0 or `{0xf9,0x80,0}` for -0.0.  However, if your application encodes arrays or maps containing floating-point numbers then it's more difficult to work around.

**Comment by winitzki** [comment]
My application constructs a large `CBORObject` that may contain `Double` values inside, some of them may be zero. I would like to be able to write the entire `CBORObject` correctly. Can I somehow inject the right code at the right place? Is there a way of defining a custom class, say `MyDouble`, and define some code so that `CBORObject.FromObject(myDouble)` will serialize correctly? I was not able to find out how to do that.

**Comment by peteroupc** [comment]
Given that the CBOR object you want to serialize may be non-trivial, I suggest it would be better for you to fix the bug you found and submit a pull request, given that you've detailed exactly where the bug's source is &mdash; which is extraordinary for a bug report.

Although that is the preferred alternative, you can also write a version of `EncodeToBytes` that make use of the low-level method `CBORObject.WriteValue` and `CBORObject.Write`.  See [`Interop.md`](https://github.com/peteroupc/CBOR/blob/master/Interop.md), from the .NET version of this library, for an idea of what the solution could look like.

**Comment by winitzki** [comment]
I have submitted a PR. https://github.com/peteroupc/CBOR-Java/pull/25

While running maven tests on my local machine I have found some issues (code that did not compile with Java 8 etc.) but did not fix them in the PR.

Thank you for pointing out `Interop.md`. However I don't see how I could implement that kind of interop in Java. The Java code base doesn't have a `CBORWriter`. Instead, the encoding methods are scattered around specific CBORObject types, and all those methods are `static` so I cannot override them. I'm not sure how
